src/components/utils/src/timer.cc

```diff
@@ -29,190 +29,129 @@
-                "Timer has finished counting. Timeout(ms): "
-                    << static_cast<uint32_t>(timeout_ms_));
-  {
-    // Task locked by own lock because from this task in callback we can
-    // call Stop of this timer and get DeadLock
-    sync_primitives::AutoLock auto_lock(task_lock_);
-    DCHECK(task_);
-    task_->run();
-  }
-  sync_primitives::AutoLock auto_lock(lock_);
-  if (is_running_) {
-    const bool stop_result = StopUnsafe();
-    DCHECK_OR_RETURN_VOID(stop_result);
-  }
-  if (repeatable_) {
-    StartUnsafe();
-  }
+  delegate_.set_stop_flag(single_shot_);
+  sync_primitives::AutoLock auto_lock(task_lock_);
+  DCHECK(task_);
+  task_->run();
+}
+
+timer::Timer::TimerDelegate::TimerDelegate(const Timer* timer)
+    : timer_(timer)
+    , timeout_(0)
+    , stop_flag_(true)
+    , state_lock_()
+    , params_lock_() {
+  DCHECK(timer_);
+}
+
+void timer::Timer::TimerDelegate::set_timeout(const Milliseconds timeout) {
+  sync_primitives::AutoLock auto_lock(params_lock_);
+  timeout_ = timeout <= 0 ? 1u : timeout;
+}
+
+timer::Milliseconds timer::Timer::TimerDelegate::timeout() const {
+  sync_primitives::AutoLock auto_lock(params_lock_);
+  return timeout_;
 }
 
-void timer::Timer::SetTimeoutUnsafe(const timer::Milliseconds timeout) {
-  timeout_ms_ = (0u != timeout) ? timeout : 1u;
+void timer::Timer::TimerDelegate::set_stop_flag(const bool stop_flag) {
+  sync_primitives::AutoLock auto_lock(params_lock_);
+  stop_flag_ = stop_flag;
 }
 
-void timer::Timer::StartUnsafe() {
-  LOG4CXX_DEBUG(logger_, "Creating posix_timer in " << name_);
-  // Create new posix timer
-  timer_ = StartPosixTimer(*this, timeout_ms_);
-  DCHECK_OR_RETURN_VOID(timer_);
-  is_running_ = true;
+bool timer::Timer::TimerDelegate::stop_flag() const {
+  sync_primitives::AutoLock auto_lock(params_lock_);
+  return stop_flag_;
 }
 
-bool timer::Timer::StopUnsafe() {
-  LOG4CXX_DEBUG(logger_, "Stopping timer  " << name_);
-  //  Destroing of posix timer
-  if (StopPosixTimer(timer_)) {
-    is_running_ = false;
-    return true;
+void timer::Timer::TimerDelegate::threadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
+  set_stop_flag(false);
+  while (!stop_flag()) {
+    const Milliseconds curr_timeout = timeout();
+    LOG4CXX_DEBUG(logger_, "Milliseconds left to wait: " << curr_timeout);
+    if (sync_primitives::ConditionalVariable::kTimeout == termination_condition_.WaitFor(auto_lock, curr_timeout)) {
+      LOG4CXX_DEBUG(logger_, "Timer has finished counting. Timeout (ms): " << curr_timeout);
+      DCHECK(timer_);
+      timer_->OnTimeout();
+    } else {
+      LOG4CXX_DEBUG(logger_, "Timeout reset force (ms): " << curr_timeout);
+    }
   }
-  return false;
+  set_timeout(0);
 }
 
-timer::Milliseconds timer::Timer::GetTimeout() const {
-  sync_primitives::AutoLock auto_lock(lock_);
-  return timeout_ms_;
+void timer::Timer::TimerDelegate::exitThreadMain() {
+  sync_primitives::AutoLock auto_lock(state_lock_);
```

Comment:
@anosach-luxoft, lokks liek it will be deadlock - because `threadMain` locks this for whole cycle and the main idea of `exitThreadMain` to stop this `main`